oximeter/db/src/native/block.rs

```diff
@@ -0,0 +1,699 @@
+impl Block {
+    /// Return an iterator over each column's data type.
+    pub fn data_types(&self) -> impl Iterator<Item = &'_ DataType> {
+        self.columns.values().map(|col| &col.data_type)
+    }
+
+    /// Return true if the provided block is empty.
+    pub fn is_empty(&self) -> bool {
+        self.n_rows == 0
+    }
+
+    /// Create an empty block with the provided column names and types
+    pub fn empty<'a>(
+        types: impl IntoIterator<Item = (&'a str, DataType)>,
+    ) -> Result<Self, Error> {
+        let mut columns = IndexMap::new();
+        let mut n_columns = 0;
+        for (name, type_) in types.into_iter() {
+            if !type_.is_supported() {
+                return Err(Error::UnsupportedDataType(type_.to_string()));
+            }
+            n_columns += 1;
+            columns.insert(name.to_string(), Column::empty(type_));
+        }
+        Ok(Self {
+            name: String::new(),
+            info: BlockInfo::default(),
+            n_columns,
+            n_rows: 0,
+            columns,
+        })
+    }
+
+    /// Concatenate this data block with another.
+    ///
+    /// An error is returned if the two blocks have different structure.
+    pub(crate) fn concat(&mut self, block: Block) -> Result<(), Error> {
+        if !self.matches_structure(&block) {
+            return Err(Error::MismatchedBlockStructure);
+        }
+        for (our_col, their_col) in
+            self.columns.values_mut().zip(block.columns.into_values())
+        {
+            our_col.concat(their_col).expect("Checked above");
+        }
+        Ok(())
+    }
+
+    fn matches_structure(&self, block: &Block) -> bool {
+        if self.n_columns != block.n_columns {
+            return false;
+        }
+        for (us, them) in self.columns.iter().zip(block.columns.iter()) {
+            if us.0 != them.0 {
+                return false;
+            }
+            if us.1.data_type != them.1.data_type {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+/// Details about the block.
+///
+/// This is only used for a few special kinds of queries. See the fields for
+/// details.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct BlockInfo {
+    /// True if this is an "overflow" block, which is the case if:
+    ///
+    /// - The data results from a query like `... GROUP BY ... WITH TOTALS`
+    /// - The number of rows exceeds the `max_rows_to_group_by` setting.
+    /// - The `group_by_overflow_mode` setting is "any".
+    ///
+    /// In that case, this block contains only the aggregated overflow data
+    /// beyond `max_rows_to_group_by`.
+    pub is_overflows: bool,
+    /// Used to optimize merges for distributed aggregation.
+    pub bucket_num: i32,
+}
+
+impl BlockInfo {
+    pub const OVERFLOW_FIELD_NUM: u64 = 1;
+    pub const BUCKET_FIELD_NUM: u64 = 2;
+}
+
+impl Default for BlockInfo {
+    fn default() -> Self {
+        Self { is_overflows: false, bucket_num: -1 }
```

Comment:
I assume `-1` means something here --- it could be nice if it was written down?

Reply:
It does mean something, but only in situations that we are very unlikely to care about. I'll add a note though, thanks.